src/index.ts

```diff
@@ -151,61 +152,67 @@ async function runTests(
-        } else {
-            const latestTypesVersion = last(typesVersions);
-            const versionPath = joinPaths(dirPath, `ts${latestTypesVersion}`);
-            const versionIndexText = await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
-            await testTypesVersion(
-                versionPath, tsVersion, tsVersion,
-                isOlderVersion, dt, versionIndexText, expectOnly, tsLocal, /*inTypesVersionDirectory*/ true);
-        }
+        await testTypesVersion(dirPath, tsVersion, tsVersion, isOlderVersion, dt, expectOnly, tsLocal, /*isLatest*/ true);
     } else {
-        await testTypesVersion(dirPath, undefined, getTsVersion(0), isOlderVersion, dt, indexText, expectOnly, undefined);
-        for (let i = 0; i < typesVersions.length; i++) {
-            const version = typesVersions[i];
-            const versionPath = joinPaths(dirPath, `ts${version}`);
-            const versionIndexText = await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
-            await testTypesVersion(
-                versionPath, version, getTsVersion(i + 1), isOlderVersion, dt, versionIndexText,
-                expectOnly, undefined, /*inTypesVersionDirectory*/ true);
-        }
-
-        function getTsVersion(i: number): TsVersion {
-            return i === typesVersions.length
-                ? TypeScriptVersion.latest
-                : assertDefined(TypeScriptVersion.previous(typesVersions[i]));
+        // For example, typesVersions of [3.2, 3.5, 3.6] will have
+        // associated ts3.2, ts3.5, ts3.6 directories, for
+        // <=3.2, <=3.5, <=3.6 respectively; the root level is for 3.7 and above.
+        // so this code needs to generate ranges [lowest-3.2, 3.3-3.5, 3.6-3.6, 3.7-latest]
+        const lows = [TypeScriptVersion.lowest, ...typesVersions.map(next)];
+        const his = [...typesVersions, TypeScriptVersion.latest];
+        assert.strictEqual(lows.length, his.length);
+        for (let i = 0; i < lows.length; i++) {
+            const low = maxVersion(minVersion, lows[i]);
+            const hi = his[i];
+            assert(
+                parseFloat(hi) >= parseFloat(low),
+                `'// Minimum TypeScript Version: ${minVersion}' in header skips ts${hi} folder.`);
+            const isLatest = hi === TypeScriptVersion.latest;
+            const versionPath = isLatest ? dirPath : joinPaths(dirPath, `ts${hi}`);
+            if (lows.length > 1) {
+                console.log("testing from", low, "to", hi, "in", versionPath);
+            }
+            await testTypesVersion(versionPath, low, hi, isOlderVersion, dt, expectOnly, undefined, isLatest);
         }
     }
 }
 
+function maxVersion(v1: TypeScriptVersion | undefined, v2: TypeScriptVersion | undefined): TypeScriptVersion;
+function maxVersion(v1: AllTypeScriptVersion | undefined, v2: AllTypeScriptVersion | undefined): AllTypeScriptVersion;
```

Comment:
They can both be undefined? Maybe the return value should be wrapped in an assertDefined?

Reply:
In practise, it's only the first one, I just didn't like having different types. But I don't want to add a pointless assertion, so I think I'll change the types.